File: sdk/javascript/src/realtime-client.ts
import { Centrifuge, type Subscription } from "centrifuge";
import { requestJson, type HttpClientOptions } from "./http";

type RealtimeClientBaseOptions = {
  baseUrl: string;
  timeoutMs?: number;
};

export type EltRealtimeClientOptions = RealtimeClientBaseOptions & (
  | {
      projectKey: string;
      projectSecret: string;
      workspaceId: number;
      getAccessToken?: never;
    }
  | {
      getAccessToken: () => Promise<string | null> | string | null;
      projectKey?: never;
      projectSecret?: never;
      workspaceId?: never;
    }
);

type EventHandler = (...args: unknown[]) => void;

export class EltRealtimeClient {
  private http: HttpClientOptions;
  private tokenPath: string;
  private centrifuge: Centrifuge | null = null;
  private subscriptions = new Map<string, Subscription>();

  constructor(options: EltRealtimeClientOptions) {
    this.http = {
      baseUrl: options.baseUrl,
      auth: "projectKey" in options && options.projectKey !== undefined
        ? {
            type: "project",
            projectKey: options.projectKey,
            projectSecret: options.projectSecret,
          }
        : { type: "jwt", getAccessToken: options.getAccessToken },
      timeoutMs: options.timeoutMs ?? 60_000,
    };
    this.tokenPath = "projectKey" in options && options.projectKey !== undefined
      ? `/api/v1/workspaces/${options.workspaceId}/notifications/realtime-token`
      : "/api/v1/notifications/realtime-token";
  }

  async connect(): Promise<void> {
    const tokenRes = await requestJson<{ token: string; ws_url: string }>(
      this.http,
      "GET",
      this.tokenPath,
    );
    this.centrifuge = new Centrifuge(tokenRes.ws_url, { token: tokenRes.token });
    this.centrifuge.connect();
  }

  subscribe(channel: string, handler: (data: unknown) => void): void {
    if (!this.centrifuge) {
      throw new Error("Not connected — call connect() first");
    }
    const sub = this.centrifuge.newSubscription(channel);
    sub.on("publication", (ctx) => handler(ctx.data));
    sub.subscribe();
    this.subscriptions.set(channel, sub);
  }

  unsubscribe(channel: string): void {
    const sub = this.subscriptions.get(channel);
    if (sub) {
      sub.unsubscribe();
      sub.removeAllListeners();
      this.subscriptions.delete(channel);
    }
  }

  on(event: "connected" | "disconnected" | "error", handler: EventHandler): void {
    if (!this.centrifuge) {
      throw new Error("Not connected — call connect() first");
    }
    this.centrifuge.on(event, handler);
  }

  disconnect(): void {
    for (const channel of [...this.subscriptions.keys()]) {
      this.unsubscribe(channel);
    }
    this.centrifuge?.disconnect();
    this.centrifuge = null;
  }
}

File: sdk/python/elt_sdk/realtime_client.py
"""Realtime subscribe client — requires optional `centrifuge-python` package."""

from __future__ import annotations

from typing import Any, Callable, Dict

import httpx

from elt_sdk.client import EltClientError


class EltRealtimeClient:
  def __init__(
      self,
      base_url: str,
      *,
      get_access_token: Callable[[], str],
      timeout: float = 60.0,
  ):
      self._base_url = base_url.rstrip("/")
      self._get_access_token = get_access_token
      self._timeout = timeout
      self._client: Any = None
      self._subscriptions: Dict[str, Any] = {}
      self._handlers: Dict[str, list[Callable[[Any], None]]] = {}

  def _auth_headers(self) -> Dict[str, str]:
      token = self._get_access_token()
      return {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}

  def _fetch_token(self) -> tuple[str, str]:
      with httpx.Client(timeout=self._timeout) as http:
          resp = http.get(
              f"{self._base_url}/api/v1/notifications/realtime-token",
              headers=self._auth_headers(),
          )
      if resp.status_code >= 400:
          raise EltClientError(resp.text, resp.status_code)
      data = resp.json()
      return data["token"], data["ws_url"]

  def connect(self) -> None:
      try:
          from centrifuge import Client  # type: ignore
      except ImportError as exc:
          raise EltClientError(
              "Install centrifuge-python for realtime: pip install 'elt-sdk[realtime]'",
              0,
          ) from exc

      token, ws_url = self._fetch_token()
      self._client = Client(ws_url, token=token)
      self._client.connect()

  def subscribe(self, channel: str, handler: Callable[[Any], None]) -> None:
      if not self._client:
          raise EltClientError("Not connected — call connect() first", 0)
      from centrifuge import SubscriptionEventHandler  # type: ignore

      class _Handler(SubscriptionEventHandler):
          def on_publication(self, event):  # noqa: ANN001
              handler(event.data)

      sub = self._client.new_subscription(channel, events=_Handler())
      sub.subscribe()
      self._subscriptions[channel] = sub
      self._handlers.setdefault(channel, []).append(handler)

  def unsubscribe(self, channel: str) -> None:
      sub = self._subscriptions.pop(channel, None)
      if sub is not None:
          sub.unsubscribe()
      self._handlers.pop(channel, None)

  def on(self, event: str, handler: Callable[..., None]) -> None:
      if not self._client:
          raise EltClientError("Not connected — call connect() first", 0)
      self._client.on(event, handler)

  def disconnect(self) -> None:
      for channel in list(self._subscriptions.keys()):
          self.unsubscribe(channel)
      if self._client is not None:
          self._client.disconnect()
          self._client = None
